internal/pkg/cli/run_local.go

```diff
@@ -401,6 +412,97 @@ func (o *runLocalOpts) getTask(ctx context.Context) (orchestrator.Task, error) {
 	return task, nil
 }
 
+func (o *runLocalOpts) prepareTask(ctx context.Context) (orchestrator.Task, error) {
+	task, err := o.getTask(ctx)
+	if err != nil {
+		return orchestrator.Task{}, fmt.Errorf("get task: %w", err)
+	}
+
+	mft, _, err := workloadManifest(&workloadManifestInput{
+		name:         o.wkldName,
+		appName:      o.appName,
+		envName:      o.envName,
+		ws:           o.ws,
+		interpolator: o.newInterpolator(o.appName, o.envName),
+		unmarshal:    o.unmarshal,
+		sess:         o.envManagerSess,
+	})
+	if err != nil {
+		return orchestrator.Task{}, err
+	}
+
+	containerURIs, err := o.buildContainerImages(mft)
+	if err != nil {
+		return orchestrator.Task{}, fmt.Errorf("build images: %w", err)
+	}
+
+	// replace built images with the local built URI
+	for name, uri := range containerURIs {
+		ctr, ok := task.Containers[name]
+		if !ok {
+			return orchestrator.Task{}, fmt.Errorf("built an image for %q, which doesn't exist in the task", name)
+		}
+
+		ctr.ImageURI = uri
+		task.Containers[name] = ctr
+	}
+
+	return task, nil
+}
+
+func (o *runLocalOpts) watchLocalFiles(watchCh chan<- interface{}, watchErrCh chan<- error, stopCh <-chan struct{}) error {
+	copilotDir := o.ws.Path()
+
+	watcher, err := o.newRecursiveWatcher()
+	if err != nil {
+		return err
+	}
+
+	if err = watcher.Add(copilotDir); err != nil {
+		return err
+	}
+
+	watcherEvents := watcher.Events()
+	watcherErrors := watcher.Errors()
+
+	var debounceCh <-chan time.Time
+
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				watcher.Close()
+				return
+			case err, ok := <-watcherErrors:
+				watchErrCh <- err
+				if !ok {
+					watcher.Close()
+					return
+				}
+			case event, ok := <-watcherEvents:
+				if !ok {
+					watcher.Close()
+					return
+				}
+
+				isHidden, err := file.IsHiddenFile(event.Name)
+				if err != nil {
+					break
+				}
+				// TODO(Aiden): implement dockerignore blacklist for update
+				// Only update on Write operations to non-hidden files
+				if event.Has(fsnotify.Write) && !isHidden {
+					debounceCh = o.newDebounceTimer()
```

Comment:
creating a bunch of timers can actually be kinda resource intensive (especially without stopping them), since each gets it's own goroutine. maybe we should use https://pkg.go.dev/time#Timer with `timer.Reset()` to reset the debounce timer